src/realm/object-store/sync/app.cpp

```diff
@@ -65,9 +65,8 @@ T as(const Bson& bson)
 template <typename T>
 T get(const BsonDocument& doc, const std::string& key)
 {
-    auto& raw = doc.entries();
```

Comment:
Accessing the underlying map directly is a very significant optimization. IndexedMap lookups are a linear find.

Reply:
A am aware of this, but as the headline says, the goal here is to actually hide the underlying implementation. I have updated the description of the PR to better explain the longer goal.

Reply:
I get the goal, but without reworking map lookup to not be awful this'll have a very negative effect on parsing data sent to us from the server. There were a few places where reading the response from the BsonDocument was taking longer than constructing the request, the network roundtrip, and parsing the response combined.

Reply:
I changed the find method to use the fast underlying map directly. It actually made the code a bit nicer.